### DISCUSSION TITLE
[OKD4.7] Install FCOS 34 kernel headers from kernel-devel via machine-config like in RHCOS

### DISCUSSION THREAD

**Discussion by devzeronull** [okd-project/okd#844] [Q&A]
Hi,

is there a working way to install Fedora CoreOS Kernel-headers through a machineconfig in OKD?

The method as suggested with Red Hat CoreOS from OpenShift, described in https://github.com/falcosecurity/falco/issues/1505, by deploying it with a machine config like:
```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name: worker-extensions
spec:
  extensions:
    - kernel-devel
```
and
```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: master
  name: master-extensions
spec:
  extensions:
    - kernel-devel
```
doesn't seem to work, the MCO stucks degraded with missing packages:
> error running rpm-ostree install --allow-inactive kernel-devel: Checking out tree fb99ff5...doneEnabled rpm-md repositories: coreos-extensionsrpm-md repo 'coreos-extensions' (cached); generated: 2021-06-29T10:22:06ZImporting rpm-md...doneerror: Packages not found: kernel-devel: exit status 1

Any productive-ready idea despite building custom images would be appreciated :)

Best regards

**Comment by vrutkovs** [comment]
You need to enable Fedora repos to allow `rpm-ostree` install additional packages

**Reply by devzeronull** [reply]
Thanks for your help!

I see that all repos have been deactivated in Machine Config Daemon Firstboot through machine-configs "00-worker" and "00-master":
```
            # Disable existing repos (if any) so that OS extensions would use
            embedded RPMs only

            ExecStartPre=-/usr/bin/sh -c "sed -i 's/enabled=1/enabled=0/'
            /etc/yum.repos.d/*.repo"
```
So, is it safe to revert that and future OKD with CoreOS updates will still complete?

Is there another common way (like a flag or declaration) to enable a repo before adding extensions not embedded than creating a new machine-config that sets "enabled=1" in the repo configuration files on filesystem-level?

Sorry, but there is really rare information on it, I would also appreciate a documentation link where I could dive deeper into this topic :)

**Reply by vrutkovs** [reply]
>So, is it safe to revert that and future OKD with CoreOS updates will still complete?

No, at least we don't test this scenario.

>Is there another common way (like a flag or declaration) to enable a repo before adding extensions not embedded than creating a new machine-config that sets "enabled=1" in the repo configuration files on filesystem-level?

I'd rather not invent a new way for this, a custom MC with repo declaration seems flexible enough

**Reply by devzeronull** [reply]
Hi,
> No, at least we don't test this scenario.

okay thanks, I see ;)

> I'd rather not invent a new way for this, a custom MC with repo declaration seems flexible enough

So you mean just adding a new repo by deploying a MC that creates a file e.g. "/etc/yum.repos.d/fcos-repositories.repo" containing:
```
[fedora]
name=Fedora $releasever - $basearch
#baseurl=http://download.example/pub/fedora/linux/releases/$releasever/Everything/$basearch/os/
metalink=https://mirrors.fedoraproject.org/metalink?repo=fedora-$releasever&arch=$basearch
enabled=1
countme=1
metadata_expire=7d
repo_gpgcheck=0
type=rpm
gpgcheck=1
gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-fedora-$releasever-$basearch
skip_if_unavailable=False

[fedora-debuginfo]
name=Fedora $releasever - $basearch - Debug
#baseurl=http://download.example/pub/fedora/linux/releases/$releasever/Everything/$basearch/debug/tree/
metalink=https://mirrors.fedoraproject.org/metalink?repo=fedora-debug-$releasever&arch=$basearch
enabled=1
metadata_expire=7d
repo_gpgcheck=0
type=rpm
gpgcheck=1
gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-fedora-$releasever-$basearch
skip_if_unavailable=False

[fedora-source]
name=Fedora $releasever - Source
#baseurl=http://download.example/pub/fedora/linux/releases/$releasever/Everything/source/tree/
metalink=https://mirrors.fedoraproject.org/metalink?repo=fedora-source-$releasever&arch=$basearch
enabled=1
metadata_expire=7d
repo_gpgcheck=0
type=rpm
gpgcheck=1
gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-fedora-$releasever-$basearch
skip_if_unavailable=False
```
 
So create it like this:
```
$ cat << EOF_OUTER > fcos-repositories-workers.yaml
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name: fcos-repositories-workers
spec:
  config:
    ignition:
      config: {}
      security:
        tls: {}
      timeouts: {}
      version: 3.2.0
    networkd: {}
    passwd: {}
    storage:
      files:
      - contents:
          source: data:text/plain;charset=utf-8;base64,$(base64 -w 0 <<< '
[fedora]
name=Fedora $releasever - $basearch
#baseurl=http://download.example/pub/fedora/linux/releases/$releasever/Everything/$basearch/os/
metalink=https://mirrors.fedoraproject.org/metalink?repo=fedora-$releasever&arch=$basearch
enabled=1
countme=1
metadata_expire=7d
repo_gpgcheck=0
type=rpm
gpgcheck=1
gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-fedora-$releasever-$basearch
skip_if_unavailable=False

[fedora-debuginfo]
name=Fedora $releasever - $basearch - Debug
#baseurl=http://download.example/pub/fedora/linux/releases/$releasever/Everything/$basearch/debug/tree/
metalink=https://mirrors.fedoraproject.org/metalink?repo=fedora-debug-$releasever&arch=$basearch
enabled=1
metadata_expire=7d
repo_gpgcheck=0
type=rpm
gpgcheck=1
gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-fedora-$releasever-$basearch
skip_if_unavailable=False

[fedora-source]
name=Fedora $releasever - Source
#baseurl=http://download.example/pub/fedora/linux/releases/$releasever/Everything/source/tree/
metalink=https://mirrors.fedoraproject.org/metalink?repo=fedora-source-$releasever&arch=$basearch
enabled=1
metadata_expire=7d
repo_gpgcheck=0
type=rpm
gpgcheck=1
gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-fedora-$releasever-$basearch
skip_if_unavailable=False
')
        mode: 420
        overwrite: true
        path: /etc/yum.repos.d/fcos-repositories.repo
  osImageURL: ""
EOF_OUTER
```

And then after this MC has been deployed create and deploy another one as described before containing:
```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name: worker-extensions
spec:
  extensions:
    - kernel-devel
```

Or what do cou mean?

Best regards

**Reply by vrutkovs** [reply]
(sorry for the delay)

Yeah, that's how we expect it to work. Is there any problem with that approach?

**Reply by achilles-git** [reply]
@vrutkovs if we enable the fedora repos by applying the above MachineConfig won't it break the OKD upgrades ???

**Reply by vrutkovs** [reply]
its quite possible - at least we test upgrades in CI with all repos disabled

**Reply by achilles-git** [reply]
OK
so is there any solution for this ???

what if we keep the extension (in this case kernel-devel) installed and remove the repo during the upgrade ?? how will it impact ???

or we can remove the machine config for the extension and repo then proceed with the upgrade, once the upgrade is completed again apply the machine config to enable the repo and install the kernel-devel package.

**Reply by yuanlinios** [reply]
The "kernel-devel" installed from fedora yum repo is 5.11.12. It does not match the one used by FCOS 5.14.9. How do you work around it?
